x-pack/legacy/plugins/uptime/server/lib/adapters/saved_objects/kibana_saved_objects_adapter.ts

```diff
@@ -4,24 +4,22 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { SavedObjectsLegacyService } from 'src/core/server';
-import { UMSavedObjectsAdapter } from './types';
+import Boom from 'boom';
 import uptimeIndexPattern from './heartbeat_index_pattern.json';
+import { UMSavedObjectsAdapter } from './types.js';
 
-export class UMKibanaSavedObjectsAdapter implements UMSavedObjectsAdapter {
-  private readonly savedObjectsClient: any;
-  constructor(savedObjects: SavedObjectsLegacyService<any>, elasticsearch: any) {
-    const { SavedObjectsClient, getSavedObjectsRepository } = savedObjects;
-    const { callWithInternalUser } = elasticsearch.getCluster('admin');
-    const internalRepository = getSavedObjectsRepository(callWithInternalUser);
-    this.savedObjectsClient = new SavedObjectsClient(internalRepository);
-  }
-
-  public async getUptimeIndexPattern(): Promise<any> {
+export const UMKibanaSavedObjectsAdapter: UMSavedObjectsAdapter = {
+  getUptimeIndexPattern: async (request: any): Promise<any> => {
+    let savedObjectsClient: any;
+    try {
+      savedObjectsClient = request.getSavedObjectsClient();
+    } catch (error) {
+      throw Boom.internal('Unable to get saved objects client');
+    }
     try {
-      return await this.savedObjectsClient.get('index-pattern', uptimeIndexPattern.id);
+      return await savedObjectsClient.get('index-pattern', uptimeIndexPattern.id);
```

Comment:
Your previous implementation was always creating the index pattern in the default space, whereas the new implementation is creating the index pattern in whatever space the user happens to be in (same for retrieval). I'm not sure what the contents of this index pattern are, but is this something you want/expect users to be able to edit? Are most uptime users aware this index pattern exists, or is it an implementation detail?

Since the end-user is the one creating the index pattern now, you will need to update your feature privileges to denote that users need the ability to write index patterns:
https://github.com/elastic/kibana/blob/0cb7035d30e3b176ef89b72dd41f51344c2977aa/x-pack/legacy/plugins/uptime/server/kibana.index.ts#L41-L68


**question**: What happens when a readonly user attempts to create this index pattern? The operation will fail, but will Uptime be resilient enough to still function for them?